### PR TITLE
3570/1

### PR DESCRIPTION
**Pull request by gro-mikhail**
## Номер
3570

## Номер задания
1

## Ссылка на видео с демо
https://youtu.be/XcIYZDSB4Tk

### CODE REVIEW

File: 3570/1/outputer_basic.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      end
+    end
+    puts gems_table
+  end
+
+  private
+
+  # :reek:TooManyStatements
+  def create_arr_for_table(gem_name, gem_info)
+    arr = []
```
Comment (stahor) [inline 3570/1/outputer_basic.rb]:
```
[
  gem_name,
  gem_info[:used_by],
  ...
]
```
And look at `values_at` method of hash

File: 3570/1/outputer_basic.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class Outputer
+  def initialize(gems_info)
+    @gems_info = gems_info
+  end
+
+  # :reek:NestedIterators
+  def output_in_table
+    gems_table = Terminal::Table.new do |table|
+      table.headings = ['gem', 'used by', 'watched by', 'stars', 'forks', 'contributors', 'issues']
```
Comment (stahor) [inline 3570/1/outputer_basic.rb]: looks like constant

File: 3570/1/outputer_basic.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      @gems_info.each do |gem_name, gem_info|
```
Comment (stahor) [inline 3570/1/outputer_basic.rb]: use map to build these rows and pass them as `rows:` key into terminal table

File: 3570/1/parser.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+require 'json'
+
+class Parser
+  def initialize(gems_list)
+    @gems_list = gems_list
+  end
+
+  def parse
+    gems_info_list = {}
+    @gems_list.each do |gem_name|
```
Comment (stahor) [inline 3570/1/parser.rb]: `each_with_object`

File: 3570/1/top_gems.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  private
+
+  def sort_gems(gems_info)
+    (gems_info.sort_by { |_gem_name, info| info[:stars] }).reverse
+  end
+
+  def get_gem_for_name(gems_info)
+    result = []
+    gems_info.each { |gem| result << gem if gem[0].include?(@gem_name) }
```
Comment (stahor) [inline 3570/1/top_gems.rb]: map

File: 3570/1/top_gems.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    result = sort_gems(gems_info)
+    result = get_gem_for_name(result) if @gem_name
+    result = get_gems_for_top(result) if @top
+    Outputer.new(result).output_in_table
+  end
+
+  private
+
+  def sort_gems(gems_info)
+    (gems_info.sort_by { |_gem_name, info| info[:stars] }).reverse
```
Comment (stahor) [inline 3570/1/top_gems.rb]: parentheses are not needed here. As well as `reverse` btw. Jus add `-` before ur block return value

File: 3570/1/top_gems.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
+require 'bundler'
+require 'optparse'
+require 'yaml'
+Bundler.require(:default)
+require_relative 'outputer_basic'
+require_relative 'parser'
+
+options = {}
```
Comment (stahor) [inline 3570/1/top_gems.rb]: global variables o_O

File: 3570/1/top_gems.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    Outputer.new(result).output_in_table
+  end
+
+  private
+
+  def sort_gems(gems_info)
+    (gems_info.sort_by { |_gem_name, info| info[:stars] }).reverse
+  end
+
+  def get_gem_for_name(gems_info)
```
Comment (stahor) [inline 3570/1/top_gems.rb]: do not modify data structures in the methods. Better create new.

File: 3570/1/top_gems.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    result = []
+    gems_info.each { |gem| result << gem if gem[0].include?(@gem_name) }
+    result
+  end
+
+  def get_gems_for_top(gems_info)
+    gems_info[0..@top - 1]
```
Comment (stahor) [inline 3570/1/top_gems.rb]: `take`?

File: 3570/1/top_gems.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+  opts.on('--file=PATH') do |path|
+    options[:gems] = YAML.load_file(path)['gems']
+  end
+end.parse!
+
+options[:gems] ||= YAML.load_file('gems_list.yml')['gems']
+
+class TopGems
+  def initialize(options)
```
Comment (stahor) [inline 3570/1/top_gems.rb]: use keyword arguments

File: 3570/1/tests/top_gems_test.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+require 'minitest/autorun'
+
+class TestParser < Minitest::Test
+  def test_app_returns_result
+    output = `ruby top_gems.rb`
+    assert_includes output, 'sinatra'
```
Comment (stahor) [inline 3570/1/tests/top_gems_test.rb]: what if I change gems list?

**Comment by gro-mikhail** [comment]
@stahor refactored code

File: 3570/1/outputer_basic.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+  def print
+    puts Terminal::Table.new(
+      headings: HEAD_TABLE,
+      rows: build_rows
+    )
+  end
+
+  private
+
+  def build_rows
```
Comment (stahor) [inline 3570/1/outputer_basic.rb]: Tip: call it `rows` or `table_rows`

**Comment by gro-mikhail** [comment]
@stahor  refactored code

File: 3570/1/gem_scraper.rb [hunk 13]
```diff
@@ -0,0 +1,53 @@
+class GemScraper
+  attr_reader :gem_name
+
+  def initialize(gem_name)
+    @gem_name = gem_name
+  end
+
+  def scrape
+    body_page = gem_page
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: do you really need this variable?

File: 3570/1/gem_scraper.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+  def format_url(url)
+    url.gsub('http:', 'https:')
+  end
+
+  def sanitize_counter(block_counter)
+    block_counter.text.strip.delete(',').to_i
+  end
+
+  def get_watched_stars_forks(info_block, value)
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: `get_`?

File: 3570/1/gem_scraper.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
+class GemScraper
+  attr_reader :gem_name
+
+  def initialize(gem_name)
+    @gem_name = gem_name
+  end
+
+  def scrape
+    body_page = gem_page
+    info_block = body_page.css('.pagehead-actions')
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: you don't need this variable

File: 3570/1/option_parse.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+class OptionParse
+  # frozen string literal
+  DEFAULT_OPTIONS = { top: nil, name: nil, file: 'gems_list.yml' }.freeze
+
+  def parse
+    options = {}
+    OptionParser.new do |opts|
+      console_values(options, opts)
+    end.parse!
+    # frozen string literal
```
Comment (stahor) [inline 3570/1/option_parse.rb]: put it at the very beginning of every file

File: 3570/1/top_gems.rb [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  private
+
+  def sort_gems(gems_info)
+    gems_info.sort_by { |_gem_name, info| -info[:stars] }
+  end
+
+  def get_gem_for_name(gems_info)
+    gems_info.select { |gem| gem if gem[0].include?(gem_name) }
```
Comment (stahor) [inline 3570/1/top_gems.rb]: O_o
`gems_info.select { |gem| gem[0].include?(gem_name) }`

Comment (stahor) [reply 3570/1/top_gems.rb]: and swap with previous line

File: 3570/1/outputer_basic.rb [hunk 18]
```diff
@@ -0,0 +1,22 @@
+class Outputer
```
Comment (stahor) [inline 3570/1/outputer_basic.rb]: the file must be called `outputer.rb`

File: 3570/1/outputer_basic.rb [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+  HEAD_TABLE = ['gem', 'used by', 'watched by', 'stars', 'forks', 'contributors', 'issues'].freeze
```
Comment (stahor) [inline 3570/1/outputer_basic.rb]: TABLE_HEADERS

File: 3570/1/gem_scraper.rb [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+  def format_url(url)
+    url.gsub('http:', 'https:')
+  end
+
+  def sanitize_counter(block_counter)
+    block_counter.text.strip.delete(',').to_i
+  end
+
+  def get_watched_stars_forks(info_block, value)
+    info_block.css("a.social-count[href$=\"#{value}\"]").text.strip.delete(',').to_i
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: can't u return all 3 fields as the hash from here?

**Comment by gro-mikhail** [comment]
@stahor  refactored

File: 3570/1/gem_scraper.rb [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+  private
+
+  def github_link
+    response = Curl.get("https://rubygems.org/api/v1/gems/#{gem_name}.json").body_str
+    gem_info = JSON.parse(response)
+    format_url(gem_info['source_code_uri'] || gem_info['homepage_uri'])
+  end
+
+  def info_block
+    gem_page.css('.pagehead-actions')
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: use memoization

File: 3570/1/gem_scraper.rb [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def gem_page
+    gem_page = Curl.get(github_link)
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: memoization

File: 3570/1/gem_scraper.rb (hunk 21 above)
Comment (gro-mikhail) [reply 3570/1/gem_scraper.rb]: swears rubocop

**Comment by gro-mikhail** [comment]
refactored code

**Comment by gro-mikhail** [comment]
@stahor refactored code

File: 3570/1/parser.rb [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+
+require_relative 'gem_scraper'
+
+class Parser
+  def initialize(gems_list)
+    @gems_list = gems_list
+  end
+
+  def parse
+    @gems_list.each_with_object([]) do |gem_name, info|
```
Comment (stahor) [inline 3570/1/parser.rb]: `map`

File: 3570/1/outputer.rb [hunk 24]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class Outputer
+  TABLE_HEADER = ['gem', 'used by', 'watched by', 'stars', 'forks', 'contributors', 'issues'].freeze
+  def initialize(gems_info)
```
Comment (stahor) [inline 3570/1/outputer.rb]: newline

**Comment by gro-mikhail** [comment]
@stahor refactored

**Comment by gro-mikhail** [comment]
@stahor again refactored

File: 3570/1/gem_scraper.rb [hunk 25]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require 'json'
+
+class GemScraper
+  attr_reader :gem_name
+
+  def initialize(gem_name)
+    @gem_name = gem_name
+    pp @gem_name
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: O_O

File: 3570/1/gem_scraper.rb [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+      name: gem_name,
+      used_by: used_by,
+      contributors: sanitize_counter(gem_page.css('a[href$="contributors"] span')),
+      issues: sanitize_counter(gem_page.css('a[href$="issues"] span.Counter'))
+    }.merge(watched_stars_forks)
+  end
+
+  private
+
+  def github_link
```
Comment (stahor) [inline 3570/1/gem_scraper.rb]: use memoization

**Comment by gro-mikhail** [comment]
@stahor task 1 completed

**Pull request merged**